<h1 style='font-size:40px'> Kubernetes Production Deployment</h1>

<h2 style='font-size:30px'> 
    The Deployment Process
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Esta seção será focada em levar a aplicação web desenvolvida ao longo do curso ao Google Cloud. 
        </li>
        <li>
            O instrutor nos adverte que o projeto incorrerá <strong>custos</strong>. Por isso, procure desligá-lo após terminar as aulas. 
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
    Creating a Google Cloud Project
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            O GCP demanda que, para cada projeto criado em seu ambiente, um workspace novo seja gerado. Esse recebe o nome de "Project".
        </li>
        <li>
            Nessa aula, criamos o Project de nossa aplicação.
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
    Kubernetes Dashboard on Google Cloud
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Explicação geral da interface gráfica do GKE.
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
     Travis Deployment Overview
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Nessa aula, é explicada a lógica do arquivo .yml a ser utilizado pelo Travis CI.
        </li>
        <li>
            De maneira geral, deveremos rodar alguns testes de confiabilidade. Se esses forem bem sucedidos, encaminharemos nossas imagens ao Docker Hub para, em seguida, criarmos os componentes do cluster.
        </li>
        <li>
            No início da execução, teremos que instalar a CLI do GCP e logar no Docker CLI. Isso porque o fluxo é rodado num servidor terceiro.
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
     Running Travis CLI in a Container
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Aula em que criamos um container com a CLI do Travis CI.
        </li>
        <li>
            Usamos a instância para encriptar o arquivo com as credenciais do Google Cloud.
        </li>
    </ul>
 </div>

<figure>
        <center style='margin-top:20px'>
            <img src='img/16_travis_container.png'>
                <figcaption> Jornada de configuração de nosso container.</figcaption>
        </center>
    </figure>

<h2 style='font-size:30px'> 
     More Google Cloud CLI Config
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            Aula em que acrescentamos mais algumas configurações na CLI do GCP que será usada pelo Travis CI.
        </li>
        <li>
            A fim de consumirmos propriamente a CLI do GCP, teremos que especificar qual projeto e zona os comandos deverão ser encaminhados.
        </li>
        <li>
            Ainda, para que possamos rodar os testes, deveremos nos logar na CLI do Docker, montando a devida imagem no Docker Hub.
        </li>
    </ul>
 </div>

<h2 style='font-size:30px'> 
     Unique Tags for Built Images
</h2>
<div> 
    <ul style='font-size:20px'> 
        <li>
            O kubectl possui um problema com atualizações imperativas de imagens. Como ele compara a imagem atual e a desejada somente com base na tag informada, a CLI se recusará em realizar a atualização, caso ambas sejam referenciadas como <i> latest</i>.
        </li>
        <li>
            Nessa situação, o programa entenderá que as imagens são as mesmas. Com isso, recomenda-se associarmos uma hash a cada nova imagem, de modo a torná-la distinguível à CLI.
        </li>
        <li>
            O instrutor nos orienta a usarmos a hash do commit no qual a versão foi configurada. Isso torna tarefas de debugging mais fáceis, inclusive.
        </li>
    </ul>
 </div>

In [3]:
! code projects/complex-gh/

In [2]:
! git add .
! git commit -m 'Aula 277'
! git push

[master 6afe7f7] Aula 275 (4:00)
 4 files changed, 40 insertions(+), 15 deletions(-)
 create mode 100644 projects/complex-gh/deploy.sh
Enumerating objects: 14, done.
Counting objects: 100% (14/14), done.
Delta compression using up to 24 threads
Compressing objects: 100% (8/8), done.
Writing objects: 100% (8/8), 1.11 KiB | 1.11 MiB/s, done.
Total 8 (delta 6), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (6/6), completed with 6 local objects.
To https://github.com/felipesveiga/docker.git
   4d08186..6afe7f7  master -> master


<p style='color:red'> Vi Aula 275 e 276; Aula 277</p>

<p style='color:red'> Visite a Aula 307, caso deseje deletar o seu cluster!</p>